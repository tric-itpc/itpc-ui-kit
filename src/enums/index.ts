export enum UIKitColors {
  black = "#000",
  green = "#4DB04D",
  grey = "#B2B7C7",
  greyLight = "#E5E5E5",
  purple = "#5C53AC",
  purpleLight = "#6F65CB",
  red = "#D42564",
  redDark = "#C9184A",
  white = "#FFF",
  yellow = "#DCB21E",
}

export enum Theme {
  DARK = "dark",
  DEFAULT = "default",
}

export enum SortType {
  ASCENDING = "ascending",
  DESCENDING = "descending",
  NONE = "none",
}

export enum NumberColumns {
  ONE = "one",
  TWO = "two",
  ZERO = "zero",
}

export enum RowAlign {
  BOTTOM = "bottom",
  MIDDLE = "middle",
  TOP = "top",
}

export enum RowJustify {
  CENTER = "center",
  END = "end",
  SPACE_AROUND = "around",
  SPACE_BETWEEN = "between",
  SPACE_EVENLY = "evenly",
  START = "start",
}

export enum TextSize {
  L = "l",
  M = "m",
  S = "s",
  XL = "xl",
  XS = "xs",
  XXL = "xxl",
  XXS = "xxs",
}

export enum TextWeight {
  BOLD = "bold",
  LIGHT = "light",
  MEDIUM = "medium",
  NORMAL = "normal",
}

export enum TextType {
  DANGER = "danger",
  DISABLED = "disabled",
  PRIMARY = "primary",
  SECONDARY = "secondary",
  SUCCESS = "success",
  WARNING = "warning",
}

export enum TextTag {
  CODE = "code",
  EM = "em",
  PARAGRAPH = "p",
  STRONG = "strong",
}

export enum AnimationTypes {
  FIXED = "FIXED",
  TRANSFORM_BY_AXIS_X = "transform-by-axis-x",
  TRANSFORM_BY_AXIS_X_Y = "transform-by-axis-x_y",
}
