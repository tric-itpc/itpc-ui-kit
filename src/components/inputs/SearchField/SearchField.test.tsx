import React from "react"

import "@testing-library/jest-dom"
import { fireEvent, render, screen } from "@testing-library/react"

import { KeyCode } from "../../../enums"
import {
  blurInput,
  changeInput,
  focusInput,
  getInputValue,
  keyDownInput,
} from "../../../lab/tests"

import { Props, SearchField } from "./index"

describe("Компонент SearchField", () => {
  const defaultProps: Props = {
    items: [
      { id: "1", value: "Apple" },
      { id: "2", value: "Banana" },
      { id: "3", value: "Cherry" },
    ],
    onChange: jest.fn(),
    placeholder: "Search...",
  }

  test("Рендерится с дефолтными пропсами (снимок)", () => {
    const { container } = render(
      <SearchField {...defaultProps} data-testid="search-field" />
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  test("Отображает placeholder", () => {
    render(<SearchField {...defaultProps} data-testid="search-field" />)
    const input = screen.getByText("Search...")
    expect(input).toBeInTheDocument()
  })

  test("Отображает иконку, если она передана", () => {
    const icon = <span>🔍</span>
    render(
      <SearchField {...defaultProps} data-testid="search-field" icon={icon} />
    )
    const iconElement = screen.getByText("🔍")
    expect(iconElement).toBeInTheDocument()
  })

  test("Открывает выпадающий список при фокусе", () => {
    render(<SearchField {...defaultProps} data-testid="search-field" />)
    const search = screen.getByTestId("search-field")
    const input = search.children[0]

    changeInput(input as HTMLElement, "ban")
    focusInput(input as HTMLElement)

    const listBox = screen.getByRole("listbox")
    expect(listBox).toBeInTheDocument()
    expect(listBox).toHaveClass("itpc-list-box_opened")
  })

  test("Закрывает выпадающий список при потере фокуса", () => {
    render(<SearchField {...defaultProps} data-testid="search-field" />)
    const search = screen.getByTestId("search-field")
    const input = search.children[0]

    changeInput(input as HTMLElement, "ban")
    focusInput(input as HTMLElement)
    blurInput(input as HTMLElement)

    const listBox = screen.queryByRole("listbox")
    expect(listBox).toBeInTheDocument()
    expect(listBox).toHaveClass("itpc-list-box_closed")
  })

  test("Фильтрует элементы при вводе текста", () => {
    render(<SearchField {...defaultProps} data-testid="search-field" />)
    const search = screen.getByTestId("search-field")
    const input = search.children[0]

    changeInput(input as HTMLElement, "Ban")

    const filteredItem = screen.getByText("Banana")
    expect(filteredItem).toBeInTheDocument()
    expect(screen.queryByText("Apple")).not.toBeInTheDocument()
  })

  test("Вызывает onChange при выборе элемента", () => {
    render(<SearchField {...defaultProps} data-testid="search-field" />)
    const search = screen.getByTestId("search-field")
    const input = search.children[0]

    focusInput(input as HTMLElement)
    changeInput(input as HTMLElement, "Ban")

    const item = screen.getByText("Banana")
    fireEvent.click(item)

    expect(defaultProps.onChange).toHaveBeenCalledWith("2")
  })

  test("Обрабатывает навигацию с клавиатуры", () => {
    render(<SearchField {...defaultProps} data-testid="search-field" />)
    const search = screen.getByTestId("search-field")
    const input = search.children[0]

    changeInput(input as HTMLElement, "A")
    focusInput(input as HTMLElement)

    const firstItem = screen.getByText("Apple")
    expect(firstItem).toHaveClass("itpc-select-item_active")

    keyDownInput(input as HTMLElement, KeyCode.ARROW_DOWN)

    const lastItem = screen.getByText("Banana")
    expect(lastItem).toHaveClass("itpc-select-item_active")

    keyDownInput(input as HTMLElement, KeyCode.ENTER)
    expect(defaultProps.onChange).toHaveBeenCalledWith("2")
  })

  test("Очищает поле при нажатии на иконку", () => {
    const icon = <span>❌</span>
    render(
      <SearchField {...defaultProps} data-testid="search-field" icon={icon} />
    )
    const search = screen.getByTestId("search-field")
    const input = search.children[0]
    changeInput(input as HTMLElement, "Ban")

    const clearIcon = screen.getByText("❌")
    fireEvent.click(clearIcon)

    expect(getInputValue(input as HTMLElement)).toEqual("")
    expect(defaultProps.onChange).toHaveBeenCalledWith("")
  })

  test("Не очищает поле, если isDisableClickIcon = true", () => {
    const onChange = jest.fn()
    const icon = <span>❌</span>
    render(
      <SearchField
        {...defaultProps}
        data-testid="search-field"
        icon={icon}
        onChange={onChange}
        isDisableClickIcon
      />
    )
    const search = screen.getByTestId("search-field")
    const input = search.children[0]
    changeInput(input as HTMLElement, "Ban")

    const clearIcon = screen.getByText("❌")
    fireEvent.click(clearIcon)

    expect(getInputValue(input as HTMLElement)).toEqual("Ban")
    expect(onChange).not.toHaveBeenCalled()
  })

  test("Вызывает fetchData при изменении значения", async () => {
    const fetchData = jest.fn()
    render(
      <SearchField
        {...defaultProps}
        data-testid="search-field"
        fetchData={fetchData}
      />
    )
    const search = screen.getByTestId("search-field")
    const input = search.children[0]
    changeInput(input as HTMLElement, "Ban")

    await new Promise((resolve) => setTimeout(resolve, 500))

    expect(fetchData).toHaveBeenCalledWith("Ban")
  })
})
