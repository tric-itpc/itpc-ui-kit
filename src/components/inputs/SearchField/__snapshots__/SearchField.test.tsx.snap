// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Компонент SearchField Рендерится с дефолтными пропсами (снимок) 1`] = `
<div
  class="itpc-search-field itpc-search-field_hover "
  data-testid="search-field"
>
  <div
    class="itpc-field"
  >
    <div
      class="itpc-input-wrap itpc-input-wrap_hover"
      style="height: 40px; max-height: 200px;"
    >
      <label
        class="itpc-placeholder"
        for="itpc-search-field"
      >
        Search...
      </label>
      <input
        autocomplete="off"
        class="itpc-input"
        id="itpc-search-field"
        name="itpc-search-field"
        type="text"
        value=""
      />
      <span
        class="itpc-error"
        title=""
      />
      <span
        class="itpc-input-icon"
      >
        <div />
      </span>
    </div>
  </div>
</div>
`;
